mumd/src/network/tcp.rs

```diff
@@ -36,9 +36,52 @@ pub enum TcpEvent {
     Disconnected, //fires when the client has disconnected from a server
 }
 
+#[derive(Clone)]
 pub enum TcpEventData<'a> {
-    Connected(&'a msgs::ServerSync),
-    Disconnected,
+    Connected(Result<&'a msgs::ServerSync, mumlib::Error>),
+    _Disconnected,
+}
+
+impl<'a> From<&TcpEventData<'a>> for TcpEvent {
+    fn from(t: &TcpEventData) -> Self {
+        match t {
+            TcpEventData::Connected(_) => TcpEvent::Connected,
+            TcpEventData::_Disconnected => TcpEvent::Disconnected,
+        }
+    }
+}
+
+struct TcpEventQueue {
+    handlers: Arc<Mutex<HashMap<TcpEvent, Vec<TcpEventCallback>>>>,
+}
+
+impl TcpEventQueue {
+    fn new() -> Self {
+        Self {
+            handlers: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    async fn register(&mut self, at: TcpEvent, callback: TcpEventCallback) {
+        self.handlers.lock().await.entry(at).or_default().push(callback);
+    }
+
+    async fn send<'a>(&mut self, data: TcpEventData<'a>) {
```

Comment:
Is there a good reason for these functions taking a `&mut self` or can they simply take a `&self`?